graph_dfs_debug/graph.py

```diff
@@ -19,29 +19,36 @@ def add_vertex(self, vertex, edges=()):
         self.vertices[vertex] = set(edges)
 
     def add_edge(self, start, end, bidirectional=True):
-        self.vertices[start].add(start)
+        if start not in self.vertices and end not in self.vertices:
+            raise ValueError("No Entries.")
+
+        self.vertices[start].add(end) # start and start? its start node to end node to create the vertex
         if bidirectional:
-            self.vertices[end].add(end)
+            self.vertices[end].add(start)
 
     def dfs(self, start, target=None):
-        x = []
-        x.append(start)
-        y = set(x)
 
-        while x:
-            z = x.pop()
-            if x == target:
+        # DFS works with stacks
+
+        stack = [start] # the stack
+
+        visited = set() # visited node
+
+        while stack:
+            current = stack.pop() # why a z?
+            if current == target:
                 break
-            x.extend(self.vertices[z])
+            visited.add(current)
+            stack.extend(self.vertices[current])
```

Comment:
You already caught this during our one-on-one but adding this here for posterity:

```py
stack.extend(self.vertices[current] - visited)
```